### PR TITLE
GCP Mailer functionality Enhancement

### PR DESCRIPTION
**Pull request by bclements**
This enhancement to the c7n_mailer tool adds gcp support. The following items were included in this PR:

- GCP Pub Sub message processing.
- Notification delivery via EmailDelivery. (Slack & Datadog integration parity coming soon in subsequent PRs)
- Basic Unit Tests
- Modification/updates to README.md to include GCP Config information.

Functionality has been tested and verified using mailtrap.io smtp testing servers. 

The process flow for GCP Mailer is as follows:

- Custodian GCP Policy sends notification to GCP PubSub Topic as normal.
- C7N_Mailer is run from the command line (Google Cloud functions are not supported at this time)
- The GCP Mailer processor retrieves all messages from the queue and processes each message.
- For each message, it will send notification to the list of "To" email addresses using SMTP infrastructure configured in the config file.

How to test this code:

1. gcloud auth login
2. Create a gcp pubsub topic(for instance: gcp-notify-test1) and add a "PULL" subscription to that topic
3. Create an example policy similar to this or use another gcp resource which as resources:

```yaml
policies:
  - name: gcp-notify-test-for-mailer
    resource: gcp.function
    actions:
      - type: notify
        template: default
        subject: 'testing the c7n mailer'
        to:
          - testuser@example.com
        transport:
          type: pubsub
          topic: 'projects/myproject/topics/mytopic'
```
4. Run python cli.py -c mailerconfig.yml --run

### CODE REVIEW

File: tools/c7n_mailer/README.md [hunk 1]
```diff
@@ -486,9 +537,6 @@ The following extra global functions are available:
 | `format_struct(struct)` | pretty print a json structure |
 | `resource_tag(resource, key)` | retrieve a tag value from a resource or return an empty string, aliased as get_resource_tag_value |
 | `format_resource(resource, resource_type)` | renders a one line summary of a resource |
-| `date_time_format(utc_str, tz_str='US/Eastern', format='%Y %b %d %H:%M %Z')`| customize rendering of an utc datetime string |
```
Comment (kapilt) [inline tools/c7n_mailer/README.md]: why is this documentation being removed?

File: tools/c7n_mailer/README.md [hunk 2]
```diff
@@ -476,8 +529,6 @@ The following variables are available when rendering templates:
 | `action` | `notify` action that generated this SQS message |
 | `policy` | policy that triggered this notify action |
 | `account` | short name of the aws account |
-| `region`  | region the policy was executing in |
```
Comment (kapilt) [inline tools/c7n_mailer/README.md]: this documentation is still valid and shouldn't be removed.

File: tools/c7n_mailer/README.md [hunk 3]
```diff
@@ -502,7 +550,7 @@ The following extra jinja filters are available:
 
 Clone the repository:
 ```
-$ git clone https://github.com/cloud-custodian/cloud-custodian
```
Comment (kapilt) [inline tools/c7n_mailer/README.md]: the urls shouldn't be changed, this looks like a bad merge of trunk into a branch wrt to the readme changes.

**Comment by kapilt** [comment]
Thanks for the pr, in general it looks good, however the readme changes  here need to be undone and the gcp docs reapplied to a trunk copy of the readme.

File: tools/c7n_mailer/README.md (hunk 3 above)
Comment (bclements) [reply tools/c7n_mailer/README.md]: The README.md has been corrected in the latest commit. I actually wrote the README.md last month using the older version before the doc sprint. I just commited the latest README.md but with the gcp content now.

**Comment by bclements** [comment]
@kapilt @thisisshi Ready for review before merge.

**Comment by bclements** [comment]
@kapilt @thisisshi Ok, maybe you guys can tell me how to resolve the codecov failures. Every other check passes but codecove patch fails.

**Comment by kapilt** [comment]
you have to write tests for the changes to get the codecov to pass

File: tools/c7n_mailer/tests/test_gcp.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        self.assertEqual(processor.receive_messages(), GCP_MESSAGES)
+        mock_receive.assert_called()
+        self.assertTrue(processor.process_message(self.compressed_message))
+        processor.ack_messages('2019-05-13T18:31:17.926Z')
+        mock_ack.assert_called_once_with('2019-05-13T18:31:17.926Z')
+
+    @patch.object(MailerGcpPubSubProcessor, 'run')
+    def test_run(self, mock_run):
+        mock_run.return_value = None
+        processor = MailerGcpPubSubProcessor(MAILER_CONFIG_1, logger)
```
Comment (kapilt) [inline tools/c7n_mailer/tests/test_gcp.py]: what is this testing?

File: tools/c7n_mailer/tests/test_gcp.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        self.compressed_message.content = base64.b64encode(
+            zlib.compress(GCP_MESSAGE.encode('utf8')))
+        self.loaded_message = json.loads(GCP_MESSAGE)
+
+    @patch.object(MailerGcpPubSubProcessor, 'receive_messages')
+    @patch.object(MailerGcpPubSubProcessor, 'process_message')
+    @patch.object(MailerGcpPubSubProcessor, 'ack_messages')
+    def test_process_pubsub_message_success(self, mock_ack, mock_process, mock_receive):
+        mock_process.return_value = True
+        mock_ack.return_value = True
```
Comment (kapilt) [inline tools/c7n_mailer/tests/test_gcp.py]: what is this testing?, your mocking every method of the class except init. none of the actual implementation is being checked.

**Comment by bclements** [comment]
> you have to write tests for the changes to get the codecov to pass, your covering off on the new transport but not some of the other changes. you can click through to see what its saying is cover or not covered in the changes

Hmm, some of those changes aren't mine, i.e. I wonder if since I merged upstream changes into my branch, it thinks that those are my changes and failing. For instance the gcp/client.py file I didn't change.

**Comment by kapilt** [comment]
most of the code that's not covered for this pr is in the new module your adding here, as noted in comments on the test they aren't actually testing the implementation because they've mocked all of it.

**Comment by kapilt** [comment]
also your branch and pr should be from your fork, not pushed to the main project repo.

**Comment by bclements** [comment]
Tests will be fixed tomorrow. I get what I gotta do now.

File: tools/c7n_mailer/tests/test_gcp.py (hunk 5 above)
Comment (bclements) [reply tools/c7n_mailer/tests/test_gcp.py]: correcting in next commit or two.

**Comment by bclements** [comment]
@kapilt I got code coverage up but it Looks like I've gotta mock up the self.client.execute_command calls to properly test some of the class methods. I'll reuse your session/client mock code from the gcp notify test in my next commit if you don't mind as it seems to do what I need.  I'll ping you if I have any questions about that snippet of mock code.

**Comment by kapilt** [comment]
alternatively mock the actual api calls not your class implementation.

**Comment by bclements** [comment]
#4077

@kapilt I think it would be wise for me to redo this once @stefangordon completes the new smtp code for azure so I can take advantage of the new abstracted smtp functionality he is also adding in.  Concur on me closing this PR?

**Comment by bclements** [comment]
Closing due to upcoming code from @logachev where he refactored smtp functionality.  Once his changes are merged, I'll resubmit a new Pr incorporating his new code into gcp functionality.

**Comment by fdeswardt** [comment]
Hey @bclements do you have the SMTP bits now from @logachev to complete this PR?  Really would love to have c7n_mailer for GCP, so let me know how I can help?
@kapilt commented on reopening this PR over in #5292 and so want to figure out how we can make it happen :)

**Comment by kapilt** [comment]
@fdeswardt smtp features from @logachev are merged (hence the conflicts), @bclements switched jobs since this pr was authored and is at an azure focused org now.

**Comment by bclements** [comment]
I actually have this code ready. I'll push it later tonight for Kapil and
community review.

**Comment by fdeswardt** [comment]
Thanks @bclements and @logachev much appreciated!

**Comment by bclements** [comment]
Apologies. I’m traveling so haven’t been able to push this. Will do this
weekend when I get back to my home office.

**Comment by fdeswardt** [comment]
Hey @bclements were you able to take a look at this merge? 🤗

**Comment by kapilt** [comment]
@bclements you may need to push to your own fork of custodian, I just noticed the branch for this pr was created erroneously on the main custodian repo.

**Comment by bclements** [comment]
Already doing so in my own repo . you can delete that branch btw.
Apologies, i thought i corrected that last year before i left c1.

-B

**Comment by fdeswardt** [comment]
Hey @bclements are you able to push your code via PR from your fork?  Really can use c7n_mailer for GCP 🙏🏻

**Pull request closed without merging**